A11y inspect

lucasradaelli opened the pull request:
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*
This change adds inspect data to the a11y bridge, which can be requested via  the command line. 

Inspect nodes are lazy computed, meaning that they are only processed when invoked, so no extra space is used during normal use.


*List which issues are fixed by this PR. You must list at least one issue.*
fxbug.dev/75100

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

lucasradaelli commented:
@dnfield  @arbreng  -- could you please review this?

Dan, in order to add inspect data, I had to add the entire SemanticsNode. My estimates say that this will not be a huge memory increase, because the heavist fields are already present, where the others are not always present (like label for example). So I think this should be fine. Since you were the one who wrote thiis first, I wanted your input to see if this is okay.

Thanks!

dnfield commented on shell/platform/fuchsia/flutter/accessibility_bridge.h:
```diff
@@ -105,28 +111,22 @@ class AccessibilityBridge
-    bool is_focusable;
-    SkRect rect;
+    flutter::SemanticsNode data;
+    // int32_t id;
+    // int32_t flags;
+    // bool is_focusable;
+    // SkRect rect;
```
commented out code (and below here)

lucasradaelli replied:
Concerning size of strings:
Note that the nodes are lazy  computed, and this is achieved by setting the callback in the a11y bridge constructor. This means that this code never runs, unless explicitly called in the command line by someone who is debugging a11y. So no change in size for there. I don't think there is a need to guard this behind a flag.

Concerning increase in memory:
We did a similar test recently for chrome, where we store the entire semantic tree in the a11y bridge, instead of only used fields. Given that chrome trees are much larger than flutter's, and there was no problem there, I am confident that this is not an issue here.

In our experience, chrome trees are two to three times bigger than flutter trees and have more data as well.

lucasradaelli replied:
sorry, I think kI answered the comment in the wrong thread -- but you will get it.. Will remove  the commented parts.

dnfield replied:
For memory size: if this works for Chrome it's probably fine.

Binary size: I'm talking about the strings getting embedded in the binary we produce from this (e.g. in flutter_runner.far). There is a lot of translation logic in the new methods that goes from an enum to a string. All of those strings will have to be available in the binary now, even though we'd never expect them to be used in release. If we `#ifdef` them away that won't be an issue.

lucasradaelli replied:
I have been thinking about this one. Our goal is to be able to say, if an user says file a bug in any product, we want to dump the semantic tree and attach to the report. This means that this needs to be in the release, right? Because otherwise this usecase would not be possible.

So we had in mind:
1. User reports a bug or asks for help
2. Developer debugging
3. crash

In all these scenarios, we would like the inspect data to be available to  be attached to reports.

dnfield replied:
Ahh, ok. If this needs to be in release though, would it need certain PII removed?

lucasradaelli replied:
Guarded the change behind a flag.

dnfield commented on shell/platform/fuchsia/flutter/accessibility_bridge.h:
```diff
@@ -105,28 +111,22 @@ class AccessibilityBridge
-    int32_t flags;
-    bool is_focusable;
-    SkRect rect;
+    flutter::SemanticsNode data;
```
This will potentially keep more data in memory than necessary.

lucasradaelli replied:
answered in the thread above.

dnfield reviewed (commented):
We should guard the inspection code behind a build flag - e.g., make sure it does not get included for release builds, since the strings will add to binary size. 

We're also retaining more data now via the entire semantics node. This probably isn't a huge deal, but it would probably be good to measure whatever impact it has on one of the lower powered Fuchsia targets we care about when accessibility services are enabled.

lucasradaelli commented:
@dnfield  -- made the changes you requested to guard this behind a flag.
@arbreng  -- this should be good for you to take a look now as well.

Also fixed the test that was failing (I changed childrenin traverssal order with children in hit test order in a for when updating the code), but the unit test caught it. Fixed.

dnfield commented:
Looks like the build is failing on an unused variable.

arbreng commented on shell/platform/fuchsia/flutter/accessibility_bridge.h:
```diff
@@ -214,6 +208,28 @@ class AccessibilityBridge
+                       inspect::Node inspect_node,
+                       inspect::Inspector* inspector) const;
+
+  AccessibilityBridge::Delegate& delegate_;
```
Thank you for re-ordering these :)

lucasradaelli commented:
the build was failing because in my machine I only buit the debug version. Built debug and release versions now, both compile.

arbreng commented on shell/platform/fuchsia/flutter/accessibility_bridge.cc:
```diff
@@ -615,4 +863,71 @@ void AccessibilityBridge::OnSemanticsModeChanged(
   delegate_.SetSemanticsEnabled(enabled);
 }
 
+void AccessibilityBridge::FillInspectTree(int32_t flutter_node_id,
```
Let's just wrap FillInspectTree and all of the FooToString methods it uses completely in #if !FLUTTER_RELEASE

The only code that calls this is already wrapped inside of a #if

Completely wrapping this function (and the others) will remove even more bytes from the release binary, and you will not be forced to maintain "degenerate" FooToString functions that return only an empty string

arbreng commented on shell/platform/fuchsia/flutter/accessibility_bridge.h:
```diff
@@ -214,6 +208,31 @@ class AccessibilityBridge
                               OnSemanticsModeChangedCallback callback) override;
 
+  // Fills the inspect tree with debug information about the semantic tree.
+  void FillInspectTree(int32_t flutter_node_id,
```
Alongside my other comment:  Wrap this in a #if !FLUTTER_RELEASE

arbreng commented on shell/platform/fuchsia/flutter/accessibility_bridge_unittest.cc:
```diff
@@ -952,4 +978,55 @@ TEST_F(AccessibilityBridgeTest, Actions) {
 }
+
+TEST_F(AccessibilityBridgeTest, InspectData) {
+#if !FLUTTER_RELEASE
```
Lets wrap the entire TEST_F inside of the #if

Otherwise, in release mode you have an empty TEST_F which passes trivially

arbreng reviewed (approved):
LGTM after #if issues are resolved

lucasradaelli commented:
done!

dnfield commented:
Windows bots are due to an infra outage that we're close to resolving.
